src/App.tsx

```diff
@@ -1,34 +1,97 @@
-/* eslint-disable max-len */
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
+import { Todo } from './types/Todo';
 import { TodoList } from './components/TodoList';
-import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
+import { TodoFilter } from './components/TodoFilter';
+import { getTodos } from './api';
 import { Loader } from './components/Loader';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [loading, setLoading] = useState(false);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [query, setQuery] = useState('');
+  const [status, setStatus] = useState<string>('all');
+  const [error, setError] = useState<string | null>(null);
+
+  const filterTodos = (
+    todosArray: Todo[],
+    searchQuery: string,
+    currentStatus: string,
+  ) => {
+    return todosArray
+      .filter(todo =>
+        todo.title.toLowerCase().includes(searchQuery.trim().toLowerCase()),
+      )
+      .filter(todo => {
+        switch (currentStatus) {
+          case 'completed':
+            return todo.completed;
+
+          case 'active':
+            return !todo.completed;
+
+          case 'all':
+          default:
+            return true;
+        }
+      });
+  };
+
+  const filteredTodos = filterTodos(todos, query, status);
+
+  useEffect(() => {
+    setLoading(true);
+    setError(null);
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => setError('Failed to load. Please try again later.'))
+      .finally(() => setLoading(false));
+  }, []);
+
   return (
     <>
       <div className="section">
         <div className="container">
           <div className="box">
             <h1 className="title">Todos:</h1>
 
+            {error && <p className="notification is-danger">{error}</p>}
+
             <div className="block">
-              <TodoFilter />
+              <TodoFilter
+                query={query}
+                setQuery={setQuery}
+                status={status}
+                setStatus={setStatus}
+              />
             </div>
 
             <div className="block">
-              <Loader />
-              <TodoList />
+              {loading ? (
+                <Loader />
+              ) : (
+                <TodoList
+                  todos={filteredTodos}
+                  selectedTodo={selectedTodo}
+                  setSelectedTodo={setSelectedTodo}
+                />
+              )}
```

Comment:
Ensure that the error message is displayed even if loading fails. Currently, the `Loader` component is shown when `loading` is true, which might hide the error message. Consider adjusting the logic to display the error message when an error occurs, regardless of the loading state.